### PR TITLE
Code coverage and additional tests

### PR DESCRIPTION
**Pull request by MurathanGoktas**
It was on our target list to implement code coverage testing to assure our code quality. #23 

I made few changes to utilize the https://mesonbuild.com/Unit-tests.html automatic code coverage testing of Meson. However there are few things to have conversation on.

1. The automatic testing of meson also tests for code coverage of third party tools such as `sdsl`. That slows down the testing. We can alternatively run `gcovr` manually in `scripts/code-coverage` and speed up the process by manually targeting the files ourselves.
2. `gcovr` produces unreliable results for headers, thus currently at `scripts/code-coverage` I only assure `.cpp` files satisfy the code coverage threshold. However some of our classes do not have `.cpp` files but have only `.hpp` file such as `mi-Bf`. It is valuable to test them all. (Update: I seperated template classes to ".hpp" and ".tpp" and it is more reliable to test now)
3. To integrate this into CI we should fix GFA2 parser and its test, or remove its tests. Otherwise test fails. (Update: Removed GFA2 module)

### CODE REVIEW

**Comment by eli-schwartz** [comment]
> The automatic testing of meson also tests for code coverage of third party tools such as `sdsl`. That slows down the testing. We can alternatively run `gcovr` manually in `scripts/code-coverage` and spped up the process by manually targeting the files ourselves.

Hmm, IIRC meson should already be passing `-e subprojects` to exclude gathering reports for your third-party dependencies.

**Review by parham-k — DISMISSED** [review]

File: wrappers/extra_common.i
Comment (parham-k) [inline wrappers/extra_common.i:65]: This does fix the compile issues for the wrappers, but even though the wrappers compiled, there was an issue when I ported the miBF tests to Python in #87. Please copy the tests from there and make sure the wrappers are functional in your branch. If not, the changes in my branch will fix them. I won't merge mine into master, which is why we should combine the changes from the two branches here.

**Review by parham-k — CHANGES_REQUESTED** [review]

File: wrappers/extra_common.i
Comment (MurathanGoktas) [reply wrappers/extra_common.i:65]: Hello @parham-k, you are right that the miBF wrapper would throw an error during run-time after merging this PR. I fixed those issues in PR request #89. To keep PRs modular I suggest handling bugs in that PR as it is not only the miBF wrapper that suffers from bugs and they are all fixed in #89.

Comment (parham-k) [reply wrappers/extra_common.i:65]: Agreed, I think we should merge these two into main for the next release to avoid broken wrappers in the main branch.

**Comment by parham-k** [comment]
Since some templates have been moved to `.tpp` files, `scripts/get-files` should be modified to find `*/*.tpp`. I'm guessing the `.tpp` files won't get copied to the installation dir here when `./compile` is executed.

**Review by parham-k — COMMENTED** [review]
Thanks @MurathanGoktas. The only thing I can think of is fixing the installation of `.tpp` files when `./compile` is called.

**Comment by MurathanGoktas** [comment]
> Thanks @MurathanGoktas. The only thing I can think of is fixing the installation of `.tpp` files when `./compile` is called.

I am not sure if its necessary to copy ".tpp" as it is included in ".hpp". Also when I changed `scripts/get-files` to include ".tpp" files it threw error, check https://github.com/mesonbuild/meson/issues/11697. What do you think @parham-k ?

**Comment by vlad0x00** [comment]
What's the advantage of using a different extension for templates? Generally, templates are simply put in .hpp files.

**Comment by MurathanGoktas** [comment]
I was getting unreliable results for ".hpp" code coverage results. I wanted to separate the implementations to another file to get reliable code coverage results and also keep the code clean by having implementations seperately. According to this Stack Excange [answer](https://softwareengineering.stackexchange.com/questions/373916/c-preferred-method-of-dealing-with-implementation-for-large-templates) I seperated the code into ".hpp" and ".tpp" files. The ".tpp" files are specifically designated for templated classes, ensuring a clear and organized structure for our codebase.

**Comment by vlad0x00** [comment]
As `meson` discussion indicated, though, `.tpp` is not widely recognized. A better solution, as somebody said in the comments of your link, is to use two `.hpp` files, one with declarations and one with implementations. Although it doesn't strike me as unreadable to have declarations at the start of the `.hpp` file and implementations as its bottom.

**Comment by MurathanGoktas** [comment]
Alright, I have updated the file names accordingly, changing ".tpp" files to "-inl.hpp" files. Additionally, I verified that gcovr does not yield accurate results when we do not separate the implementations. It would be beneficial to develop a troubleshooting guide in the future to eliminate any ambiguities.

**Comment by MurathanGoktas** [comment]
Currently `clang-tidy` doesn't like that I have to `.hpp`s for a class. As far as I can see there are 3 options:

1. use` /NOLINT` or some other thing suppress those.
2. go back to ".tpp" for implementations. Possible issue:
> Thanks @MurathanGoktas. The only thing I can think of is fixing the installation of `.tpp` files when `./compile` is called.
3. combine ".hpp"s into one as it used to and have unreliable code coverage results for mi-BF and cBF classes. And exclude them from coverage assurance tests.

I am expecting a suggestion from reviewers otherwise I am tending to follow the second option.

An example of clang-tidy errors:
```
error: too many errors emitted, stopping now [clang-diagnostic-error]
/projects/btl/tgoktas/cc_btllib/btllib/include/btllib/counting_bloom_filter-inl.hpp:28:32: error: redefinition of 'CountingBloomFilter<T>' [clang-diagnostic-error]
inline CountingBloomFilter<T>::CountingBloomFilter(size_t bytes,
                               ^
/projects/btl/tgoktas/cc_btllib/btllib/include/btllib/counting_bloom_filter-inl.hpp:28:32: note: previous definition is here
```

**Comment by vlad0x00** [comment]
That error doesn't seem sensible. The redefinition is pointing to the same line. Does clang-tidy report the error exactly like that?

Also, redefinitions of functions should not happen. In fact, compiler should say so, not just clang-tidy. There's something fishy going on here. Spreading a class across multiple files should not be an issue.

**Comment by MurathanGoktas** [comment]
It was a fishy issue indeed, thanks @vlad0x00. This branch should be good to merge after resolving the conflicts.

**Review by parham-k — APPROVED** [review]
Great! Feel free to merge whenever the conflicts are resolved.

**Pull request merged**
